stream xml encoder

DerNasenach opened the pull request:
Used parser to create tree, recursively translated tree to xml

Commit:
started work on xmlEncoder + corresponding UnitTest

Commit:
started work on xmlEncoder + corresponding UnitTest

implemented Token translation

use parser to setup tree, encode tree to xml recursively

refactored Constructor, commented, added alternative testcase for multiple Attributes, minor changes

incomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributes

GABVML-34 WIP - escaped special characters, refactored to use io.writer, unit test for stream encoding

GABVML-34: finished refactoring parser to visitor pattern. started debugging.

Commit:
implemented Token translation

Commit:
Merge remote-tracking branch 'upstream/main' into main

Commit:
use parser to setup tree, encode tree to xml recursively

haselkern commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,197 @@
+	postfix  string
+}
+
+func NewEncoderFromString(text string) XMLEncoder {
```
Ich glaube hier wäre es gut, wenn du nur einen Konstruktor hättest, der sich an denen vom Lexer und Parser orientiert:
```go
func NewParser(filename string, r io.Reader) *Parser {...}
```
Wenn ich das mit einem String aufrufen möchte, kann ich als Aufrufender meinen String mit `strings.NewReader` trotzdem einfach übergeben.

DerNasenach replied:
So übernommen

Commit:
refactored Constructor, commented, added alternative testcase for multiple Attributes, minor changes

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,206 @@
+
+		}
+	}
+	/*for{
```
auskommentierter Quelltext kann weg. Dafür gibt es ja das Git

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,206 @@
+// given by its root Element to the corresponding XML.
+func encodeRek(root parser2.TreeNode) (string, error) {
+	if root.IsComment() {
+		return "<!-- " + *root.Comment + " -->", nil
```
hier scheint mir überall das Escaping zu fehlen, zudem ist das Konkatenieren von Strings sehr ineffizient. Auf einen kompletten Buffer sollte grundsätzlich verzichtet werden, da eine Streamverarbeitung gemacht werden soll, d.h. du bekommst erstmal nur einen io.Writer

DerNasenach replied:
Heißt, ich soll nur die Konkatenation über den io.Writer machen und einen fertigen String zurückgeben? Oder soll ich den Writer nur mit Inhalt füllen und direkt zurückgeben? Escaping baue ich ein, Streamverarbeitung hatte Lars mir erklärt, das schreibe ich dahingehend um.

DerNasenach replied:
Oder ist gemeint, dass ich ein Objekt übergeben bekomme, welches io.Writer implementiert, und meine Ausgabe auf diesem Objekt realisieren soll?

torbenschinke replied:
> Heißt, ich soll nur die Konkatenation über den io.Writer machen und einen fertigen String zurückgeben?
verstehe nicht was du meinst, klingt aber unpraktisch. Ein String ist immer teuer und erfordert idR. überall kopien und zusätzliche Allocations. Du solltest grundsätzlich von außen nur einen io.Writer akzeptieren. Intern kannst du dir ja noch einen bufio.Writer erzeugen, aber das flush nicht vergessen.

torbenschinke commented on stream-xml-encoder/encoder_test.go:
```diff
@@ -0,0 +1,236 @@
+		},
+		{
+			// might fail, as the sequence in which Attributes are added to an identifier
+			// depends on the sequence they lie in the Nodes' Attributes-Slice
```
der Kommentar ist etwas schwer verständlich

Commit:
incomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributesincomplete - XMLEncoder to stream. Encoded Identifiers + possible Attributes

Commit:
GABVML-34 WIP - escaped special characters, refactored to use io.writer, unit test for stream encoding

Commit:
GABVML-34: finished refactoring parser to visitor pattern. started debugging.

torbenschinke commented:
du hast Konflikte. Probiere mal eine saubere Historie mit "rebasen" zu bekommen. Du kannst vorher auch mal Markus fragen.

torbenschinke commented:
benutze das WIP-Flag, damit ich weiß, ob ich das wirklich zum Mergen ist oder nur für ein Review

Commit:
GABVML-34: fixed rebase errors

Commit:
GABVML-34: continued debugging parser

Commit:
GABVML-34 debug parser refactor

torbenschinke commented on .vscode/launch.json:
```diff
@@ -0,0 +1,15 @@
+{
```
diese Datei sollte nicht eingecheckt werden

torbenschinke commented on go.mod:
```diff
@@ -2,4 +2,8 @@ module github.com/golangee/tadl
 
-require github.com/r3labs/diff/v2 v2.13.1
+require (
+	github.com/davecgh/go-spew v1.1.1 // indirect
```
wozu sind diese ganzen Abhängigkeiten? (go mod tidy)

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,324 @@
+	writer     io.Writer
+	buffWriter *bufio.Writer
+
+	tadlText string
```
was sind das für strings?

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+// Tokenize creates a Slice of consecutive Tokens, representing the tadl-input syntax
+func (x *XMLEncoder) Tokenize() error {
```
irgendwie macht das keinen Sinn: hier wird ja alles gepuffert. Die Anforderung war doch, dass alles als Stream verarbeitet werden soll.

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,324 @@
+
+// encodeRek recursively translates the syntax tree
+// given by its root Element to the corresponding XML.
+func encodeRek(root parser.TreeNode) (string, error) {
```
Escaping fehlt

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,324 @@
+
+// Next returns the next XML Translation
+// to the corresponding TADL token in the input stream.
+func (x *XMLEncoder) Next() error {
```
scheint mir als Public-Methode keinen Sinn zu machen. Private?

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,324 @@
+func (x *XMLEncoder) EncodeToXML() (string, error) {
+	var err error
+	var output string
+	var pars = parser.NewParser("test", bytes.NewBuffer([]byte(x.tadlText)))
```
wieso "test"?

torbenschinke commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,324 @@
+				return "", err
+			}
+
+			outString += text
```
hier wird komplett gepuffert. Das war nicht das Ziel. Lass uns noch mal zu dritt über das Thema "Streaming" reden und worauf es dabei ankommt.

Commit:
GABVML-34 finished debugging G1 parsing. Start debugging G2 parsing

Commit:
Merge remote-tracking branch 'upstream/main' into HEAD

DerNasenach commented:
@torbenschinke Der Encoder ist momentan noch nicht richtig implementiert, verwendet derzeit noch den vollständig geparsten Baum. Für die Implementierung des Streaming müsste man quasi den Parser nachbauen, nur an den Stellen, an den der Baum konstruiert wird, stattdessen ein Streamencoding verwenden. (Jedenfalls fiel mir keine andere Möglichkeit ein, da ich den Syntaxbaum ja ebenfalls traversieren muss, dies im Parser allerdings schon implementiert ist) Daher hatte Lars die Idee, den Parser zu entkoppeln und nach dem Visitor Pattern umzugestalten. So hätte ich einen Visitor, welcher das traversieren des Baumes übernimmt, so wie den Parser hinter einem entsprechenden Interface, welcher den Baum als Objekt aufbaut. Dann ließe sich der Parser einfach durch einen Encoder ersetzen, welcher das gleiche Interface bedient, und an den entsprechenden stellen stattdessen das übersetzte Element auf seinen Writer schreibt. Daran abreite ich derzeit. Oder gibt es eine andere Möglichkeit, die Eingabe richtig zu encoden, ohne den daraus resultierenden Baum traversieren zu müssen?

Commit:
GABVML-34 Improvements based on Lars' feedback: No Children on Comment- and Textnodes. refactor parserAttributes() method to call interface, moving buffered Attributes from visitor to parser.

redo active Tree switch. Now SwitchActiveTree() switches the root- and parent-pointers with the rootForward- and parentForward-pointers. Debugging

Commit:
merge with changes in G2parser (arrow parsing) and adapting to Visitor Pattern.

Commit:
GABVML-34 integrated return Arrow parsing from recent changes, finished debugging parsing

Commit:
GABVML-34 finished refactoring of parser to visitor-pattern. small changes and cleanup

Commit:
GABVML-34 started refactoring Encoder to match visitor Pattern

Commit:
GABVML-34 finished refactoring encoder to match visitable interface, implemented TreeNode inheritance, started encoding to xml and writing to io.Writer, started testing/debugging of encoder

Commit:
GABVML-34 finished debugging encoder/visitor, started expanding TestSuite and removing redundancies

Commit:
GABVLM-34 commented; ordered Attributes alphabetically when writing, edited Testcases correspondingly; test for streaming capability: first attempt

haselkern commented on stream-xml-encoder/encoder.go:
```diff
@@ -0,0 +1,451 @@
+)
+
+const (
+	whitespace = "\x20"
```
Hier besser einfach die Zeichen nehmen.

Commit:
GABVML-34 added TODOs, files to .gitignore

Commit:
GABVML-34 refactor of encoder - use stack instead of Tree, reduce redundacies. exchange AttributeMap with AttributeList, providing order-sensitivity. Started debugging

Commit:
GABVML-34 AttributeMap -> AttributeList refactor finished, encoder - tree -> stack refactor finished, debugging parser,
commented visitable interfacce, moved buffering Position to visitor

Commit:
GABVML-34 finished refactoring encoder to stack + debugging. refactor AttributeList Array -> double linked List - increased performance. introduced error handling to visitable interface + parser + encoder. minor refactors (method permissions, remove unused methods etc.)

Commit:
GABVML-34 fixed Comma-bug in visitor, introduced BlockType encoding, expanded TestSuite, used Stringbuilder for escaping doublequotes, linting

Commit:
GABVML-34 linted

Commit:
GABVML-34 redo forwarding Nodes to use stringBuilder - buffer. Add test for escaping double-quotes. commenting. minor refactors and cleaning.

haselkern commented on .gitignore:
```diff
@@ -16,4 +16,13 @@
+.antlr
+
+# VScode debug files
+*/launch.json
```
Am besten den ganzen Ordner ausschließen:

    .vsode/

DerNasenach replied:
übernommen

haselkern commented on .gitignore:
```diff
@@ -16,4 +16,13 @@
+*/launch.json
+
+# MacOS autogenerated files
+*/.DS_Store
```
Hier könntest du einfach nur `.DS_Store` schreiben, dann werden nicht nur Dateien in Unterverzeichnissen ignoriert.

DerNasenach replied:
übernommen

haselkern commented on encoder/encoder.go:
```diff
@@ -0,0 +1,452 @@
+		visitor:    *parser.NewVisitor(nil, token.NewLexer(filename, r)),
+		buffWriter: bufio.NewWriterSize(w, buffsize),
+	}
+	encoder.visitor.SetVisitable(&encoder)
```
Das kompiliert bei mir nicht 🙈 
```
./encoder.go:131:31: cannot use &encoder (type *Encoder) as type parser.Visitable in argument to encoder.visitor.SetVisitable:
	*Encoder does not implement parser.Visitable (missing AddForwardAttribute method)
```

DerNasenach replied:
übernommen

DerNasenach replied:
Da hat er wohl die Methodennamen-refactors nicht ganz richtig übernommen, sollte nun kompilierbar sein.

haselkern commented:
Am besten mergst du auch nochmal main von hier rein, um die Konflikte aufzulösen.

haselkern commented on encoder/encoder.go:
```diff
@@ -0,0 +1,452 @@
+func escapeDoubleQuotes(in string) string {
+	var out strings.Builder
+	for _, c := range in {
+		if c == '"' {
```
Hier fehlt noch der Fall, dass ein '\' auftritt, welches zu "\\" escaped werden muss.

DerNasenach replied:
Noch unklar. Wann tritt das denn auf, bzw. warum?

haselkern replied:
Wenn ich die Zeichenkette `hallo\` codieren möchte ohne den `\` zu escapen, wird daraus `"hallo\"`. Jetzt denkst du aber, dass am Ende des Strings das `"` escaped wurde, also Teil des Strings ist, und ein abschließendes `"` fehlt. Daher muss `hallo\` zu `"hallo\\"` werden. Dann ist klar, dass `\\` ein einzelner Backslash ist und nicht das abschließende `"` beeinflusst.

haselkern commented:
Kann es sein, dass mit git was schiefgelaufen ist? Ich sehe hier den Ordner `encoder` und `stream-xml-encoder`, die sich beide verdächtig ähnlich aussehen 😅

Commit:
Merge branch 'main' into HEAD

DerNasenach commented:
@haselkern Hab nochmal neu rebased

Commit:
GABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debugGABVML-34 fix trailing Backspace in lexer, debug
